File: Webfuel.Api/Api/Comms/TriageTemplateApi.cs
using MediatR;
using Microsoft.AspNetCore.Mvc;
using Webfuel.Common;
using Webfuel.Domain;
using Webfuel.Domain.StaticData;

namespace Webfuel.App
{
    [ApiService]
    [ApiDataSource]
    public static class TriageTemplateApi
    {
        public static void RegisterEndpoints(IEndpointRouteBuilder app)
        {
            // Commands
            
            app.MapPost("api/triage-template", Create)
                .RequireClaim(c => c.CanEditStaticData);
            
            app.MapPut("api/triage-template", Update)
                .RequireClaim(c => c.CanEditStaticData);

            app.MapPut("api/triage-template/sort", Sort)
                .RequireClaim(c => c.CanEditStaticData);

            app.MapDelete("api/triage-template/{id:guid}", Delete)
                .RequireClaim(c => c.CanEditStaticData);

            app.MapPut("api/triage-template/generate-email", GenerateEmail)
                .RequireIdentity();

            // Querys

            app.MapPost("api/triage-template/query", Query)
                .RequireIdentity();

            app.MapGet("api/triage-template/{id:guid}", Get)
                .RequireIdentity();
        }
        
        public static Task<TriageTemplate> Create([FromBody] CreateTriageTemplate command, IMediator mediator)
        {
            return mediator.Send(command);
        }
        
        public static Task<TriageTemplate> Update([FromBody] UpdateTriageTemplate command, IMediator mediator)
        {
            return mediator.Send(command);
        }

        public static Task Sort([FromBody] SortTriageTemplate command, IMediator mediator)
        {
            return mediator.Send(command);
        }

        public static Task Delete(Guid id, IMediator mediator)
        {
            return mediator.Send(new DeleteTriageTemplate { Id = id });
        }

        public static Task<SendEmailRequest> GenerateEmail([FromBody] GenerateTriageTemplateEmail command, IMediator mediator)
        {
            return mediator.Send(command);
        }

        public static Task<QueryResult<TriageTemplate>> Query([FromBody] QueryTriageTemplate command, IMediator mediator)
        {
            return mediator.Send(command);
        }

        public static async Task<TriageTemplate?> Get(Guid id, IMediator mediator)
        {
            return await mediator.Send(new GetTriageTemplate { Id = id });
        }
    }
}

File: Webfuel.App/ClientApp/src/app/api/triage-template.api.ts
import { EventEmitter, Injectable, inject } from '@angular/core';
import { Observable, tap } from 'rxjs';
import { ApiService, ApiOptions } from '../core/api.service';
import { ActivatedRouteSnapshot, ResolveFn, RouterStateSnapshot } from '@angular/router';
import { IDataSource } from 'shared/common/data-source';
import { CreateTriageTemplate, TriageTemplate, UpdateTriageTemplate, SortTriageTemplate, GenerateTriageTemplateEmail, SendEmailRequest, QueryTriageTemplate, QueryResult } from './api.types';

@Injectable()
export class TriageTemplateApi implements IDataSource<TriageTemplate, QueryTriageTemplate, CreateTriageTemplate, UpdateTriageTemplate> {
    constructor(private apiService: ApiService) { }
    
    public create (body: CreateTriageTemplate, options?: ApiOptions): Observable<TriageTemplate> {
        return this.apiService.request<CreateTriageTemplate, TriageTemplate>("POST", "api/triage-template", body, options).pipe(tap(_ => this.changed.emit()));
    }
    
    public update (body: UpdateTriageTemplate, options?: ApiOptions): Observable<TriageTemplate> {
        return this.apiService.request<UpdateTriageTemplate, TriageTemplate>("PUT", "api/triage-template", body, options).pipe(tap(_ => this.changed.emit()));
    }
    
    public sort (body: SortTriageTemplate, options?: ApiOptions): Observable<any> {
        return this.apiService.request<SortTriageTemplate, any>("PUT", "api/triage-template/sort", body, options).pipe(tap(_ => this.changed.emit()));
    }
    
    public delete (params: { id: string }, options?: ApiOptions): Observable<any> {
        return this.apiService.request<undefined, any>("DELETE", "api/triage-template/" + params.id + "", undefined, options).pipe(tap(_ => this.changed.emit()));
    }
    
    public generateEmail (body: GenerateTriageTemplateEmail, options?: ApiOptions): Observable<SendEmailRequest> {
        return this.apiService.request<GenerateTriageTemplateEmail, SendEmailRequest>("PUT", "api/triage-template/generate-email", body, options);
    }
    
    public query (body: QueryTriageTemplate, options?: ApiOptions): Observable<QueryResult<TriageTemplate>> {
        return this.apiService.request<QueryTriageTemplate, QueryResult<TriageTemplate>>("POST", "api/triage-template/query", body, options);
    }
    
    public get (params: { id: string }, options?: ApiOptions): Observable<TriageTemplate> {
        return this.apiService.request<undefined, TriageTemplate>("GET", "api/triage-template/" + params.id + "", undefined, options);
    }
    
    changed = new EventEmitter<any>();
    
    // Resolvers
    
    static triageTemplateResolver(param: string): ResolveFn<TriageTemplate> {
        return (route: ActivatedRouteSnapshot, state: RouterStateSnapshot): Observable<TriageTemplate> => {
            return inject(TriageTemplateApi).get({id: route.paramMap.get(param)! });
        };
    }
}

